compiler/codegen/OMRCodeGenerator.hpp

```diff
@@ -324,6 +324,8 @@ class OMR_EXTENSIBLE CodeGenerator
    TR_BitVector *getLiveLocals() {return _liveLocals;}
    TR_BitVector *setLiveLocals(TR_BitVector *v) {return (_liveLocals = v);}
 
+   bool allowGuardMerging() { return false; }
+
```

Comment:
I don't quite follow this API. It is completely unused in OMR as far as I can tell since no one overrides it to return `true`. Therefore I don't think this API needs to exist in OMR. Could we completely eliminate it, and migrate the current OMR implementation of `supportsMergingGuards` override in x86 and Z codegens to OpenJ9, since that API returns `false` in OMR always currently.

Reply:
@fjeremic: I'll move `allowGuardMerging()` down to OpenJ9. 

There is a difference across the codegens for the `supportsMergingGuard()` routine in OMR
For x86, the implementation is:
```
   return self()->getSupportsVirtualGuardNOPing() &&
          self()->comp()->performVirtualGuardNOPing() &&
          self()->allowGuardMerging();
```
This routine will always return false as you mentioned above.

For Z, the implementation is:
```
return self()->getSupportsVirtualGuardNOPing() &&
          self()->comp()->performVirtualGuardNOPing() &&
          !self()->comp()->compileRelocatableCode();
```
This will return true in standalone OMR for many cases.

I suspect in practice the Z routine can work for both platforms and we should be able to common that one up. Does that sound correct? 

If so, then the override in OpenJ9 can be placed inside the common codegen as well with the following implementation:
```
 return self()->getSupportsVirtualGuardNOPing() &&
          self()->comp()->performVirtualGuardNOPing() &&
          !self()->allowGuardMerging();
```

Reply:
There is a `!` in front of `allowGuardMerging` which I don't think looks correct. Otherwise yes we should remove `allowGuardMerging` from OMR and push up the `supportsMergingGuards` into OpenJ9. I'm not sure about putting it in the common codegen because it looks like Power and AArch64 don't implement that API. Perhaps you can open up a PR in OpenJ9 with the common codegen change and tag Andrew in a review comment to give the blessing. I'm not sure why we don't support guard merging on Power. Perhaps it is just an oversight?

Reply:
d556110

Reply:
At the moment allowGuardMerging() is doing the same thing as compileRelocatableCode because in OpenJ9 the base implementation says true and the AOT frontend override returns false. We should just simplify on the compileRelocatableCode query and simplify the whole thing - I don't see why OpenJ9 even needs an override unless some sees something more subtle. I am fine with communing the query since it is generally useful.

AArch64 hasn't gotten to support thing this yet. Not sure about POWER - they may have patching restrictions that prevent it would be my only thought.

Reply:
@andrewcraik Yes, replacing `allowGuardMerging()` with `compileRelocatableCode()` will yield the same behaviour. This will work for JITServer (it only needs to be disabled for AOT remote compilations) as well. The only reason why I opened these PRs was because there was an inconsistency between the x86 and Z codgens, and using the `allowGuardMerging()` solution seemed like the correct approach because I thought it would make it easier in the future to enable/disable support for the different front ends in the future.

However if it's simpler to just remove the `allowGuardMerging()` and use `!compileRelocatableCode()` query instead, then I'll revert the changes.

@mpirvu FYI

Reply:
Well I am happy to hear the opinion of others, but the allowGuardMerging design does allow for much more complex control patterns, but if we don't need them then it is over engineered and it is easier to follow the simpler implementation. @fjeremic thoughts? I think keeping it simple is probably best and we can common the code up. The more complex solution can always be added back if we find we need it. I think it may have originally been more complex, but as limitations have been removed things have gotten simpler.

Reply:
I agree in keeping things simple and just using `compileRelocatableCode()`. If the need for such complexity arises in the future we can introduce it then.

Reply:
This PR now refactors `supportsMergingGuards()` to the common codegen based on the above discussion. @andrewcraik As discussed offline, the support is enabled by default and the Power/AARCH64 codgens now opt-in to disable this feature.